builder/dockerignore/dockerignore.go

```diff
@@ -33,3 +36,75 @@ func ReadAll(reader io.ReadCloser) ([]string, error) {
 	}
 	return excludes, nil
 }
+
+// ReadAllRecursive takes a base directory and a starting directory,
+// reads all .dockerignore files in subdirectories and returns the list of file
+// patterns to ignore. Note this will trim whitespace from each line as well
+// as use GO's "clean" func to get the shortest/cleanest path for each.
+func ReadAllRecursive(baseDir string, directory string) ([]string, error) {
+
+	var excludes []string
+
+	directory = filepath.Clean(directory)
+	directory = filepath.ToSlash(directory)
+
+	dockerIgnFName := filepath.Join(directory, ".dockerignore")
+	reader, err := os.Open(dockerIgnFName)
+	// Note that a missing .dockerignore file isn't treated as an error
+	if err != nil {
+		// If file does not exist, ignore, and continue recursing as needed
+		// If file exists, and there is an error reading it, fail.
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		// If reader is nil, it means that something odd happened:
+		// lack of permissions, etc.
+		if reader == nil {
+			return nil, nil
+		}
+
+		ignFiles, err := ReadAll(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		relDirectory, err := filepath.Rel(baseDir, directory)
+		if err != nil {
+			return nil, err
+		}
+
+		for i := range ignFiles {
+			isException := ignFiles[i][0] == '!'
+			if isException && len(ignFiles[i]) > 1 {
+				ignFiles[i] = "!" + filepath.Clean(filepath.Join(relDirectory, ignFiles[i][1:len(ignFiles[i])]))
+			} else {
+				ignFiles[i] = filepath.Clean(filepath.Join(relDirectory, ignFiles[i]))
```

Comment:
To clarify, do you mean that we should disallow those kinds of rules? 

Otherwise, i think `filepath.Join` would allow it as per normal?